### PR TITLE
Team broker client UI

### PR DESCRIPTION
**Pull request by hardillb**
fixes https://github.com/FlowFuse/flowfuse/issues/4433
## Description

Starts to add the UI for adding Broker Clients

Still needs a lot of work

#1350 

## Related Issue(s)

https://github.com/FlowFuse/flowfuse/issues/4433

## Checklist

 - [ ] I have read the [contribution guidelines](https://github.com/FlowFuse/flowfuse/blob/main/CONTRIBUTING.md)
 - [ ] Suitable unit/system level tests have been added and they pass 
 - [ ] Documentation has been updated
    - [ ] Upgrade instructions
    - [ ] Configuration details
    - [ ] Concepts
 - [ ] Changes `flowforge.yml`?
    - [ ] Issue/PR raised on `FlowFuse/helm` to update ConfigMap Template
    - [ ] Issue/PR raised on `FlowFuse/CloudProject` to update values for Staging/Production

## Labels

 - [ ] Includes a DB migration? -> add the `area:migration` label

### CODE REVIEW

**Comment by hardillb** [comment]
@joepavitt this is functional enough for me to test things, but needs a lot of love.

I also need to write the backend API for edits to the ACL and to allow password changes. Will get those done tomorrow.

**Comment by joepavitt** [comment]
@cstns how you looking on Visual History? Need me to jump in here, or could you finish both, visual history _and_ this by Thursday next week?

**Comment by cstns** [comment]
I think I could tackle both! My best guess is that the Visual History ui will be done by the end on this week and can jump to this one afterwards.

**Comment by hardillb** [comment]
Couple of points

- create/edit/delete should only be available to team owner
- dashboard only users should not have access

**Comment by hardillb** [comment]
Also when displaying the username, it should have `@[teamHashId]` added to the end.

e.g. is username `alice`  is created in a team with a hash id of `1234abc` then the username should be displayed as `alice@1234abc`

Might be worth adding another dialog shown after creating a user that shows the username (and client id with the same value) and the topics.

**Comment by hardillb** [comment]
Quick note from testing latest checked in code.

- The Create client button need to only be available to Team Owners
- Likewise delete option when implemented

**Comment by cstns** [comment]
@Yndira-E, please give us a hand with the empty state icon and broker pictograms! 

I created https://github.com/FlowFuse/flowfuse/issues/4680

**Comment by hardillb** [comment]
Editing a client is not populating the existing rules into the dialog

image


image

**Comment by hardillb** [comment]
~~Sorry to drop this on you so late, but how hard would it be to add an extra column? called client id and shows the same as the username?~~

Actually scratch that, can we change the username field to had the header "Username/ClientId"

File: frontend/src/pages/team/Library/TeamLibrary.vue [hunk 1]
```diff
@@ -1,7 +1,7 @@
 <template>
     <!-- set mb-14 (~56px) on the form to permit access to kebab actions where hubspot chat covers it -->
     <div class="ff-team-library mb-14">
-        <div class="breadcrumbs-wrapper">
+        <div v-if="isSharedLibraryFeatureEnabled" class="breadcrumbs-wrapper">
```
Comment (hardillb) [inline frontend/src/pages/team/Library/TeamLibrary.vue:4]: I assume this is a fix for something you spotted while working on this?

**Review by hardillb — COMMENTED** [review]
I've flagged a few things

**Comment by hardillb** [comment]
We should be able to remove all, just not the last one.

As it is we can never remove the first rule.

File: frontend/src/pages/team/Broker/index.vue [hunk 2]
```diff
@@ -0,0 +1,375 @@
...
+                            >
+                                <template #icon-left>
+                                    <PlusSmIcon />
+                                </template>
+                                Create Client
+                            </ff-button>
+                        </div>
+                        <div class="clients-wrapper">
+                            <div class="header grid grid-cols-6 gap-4 font-bold">
+                                <span class="username">Username</span>
```
Comment (hardillb) [inline frontend/src/pages/team/Broker/index.vue:65]:
```suggestion
                                <span class="username">Username/ClientId</span>
```

**Comment by cstns** [comment]
> We should be able to remove all, just not the last one.
> 
> As it is we can never remove the first rule.
> 
> Screenshot from 2024-10-23 13-24-22

Ill come up with a more comprehensive solution as a follow up I don't think i can come up with a solution that would preserve the 1 acl rule mandatory to a client in a very short time

**Review by joepavitt — CHANGES_REQUESTED** [review]
Still have "Hello World" in the Empty State - otherwise, looking good.

**Comment by hardillb** [comment]
fixed the empty state

image

**Comment by cstns** [comment]
> fixed the empty state
> 
> image

Was just about to do so, spotted it whilst writing e2e tests. 

I took the liberty to make the logo smaller and add some accompanying text

File: test/e2e/frontend/cypress/tests/instances/version-history/snapshots.spec.js [hunk 3]
```diff
@@ -48,11 +49,14 @@ describe('FlowForge - Instance Snapshots', () => {
         cy.visit(`/instance/${projectId}/version-history/snapshots`)
         cy.wait('@snapshotData')
 
+        // eslint-disable-next-line cypress/require-data-selectors
```
Comment (joepavitt) [inline test/e2e/frontend/cypress/tests/instances/version-history/snapshots.spec.js:52]: This shouldn't be needed as you _are_ using a `data` selector?

File: test/e2e/frontend/cypress/tests/instances/version-history/snapshots.spec.js [hunk 4]
```diff
@@ -48,11 +49,14 @@ describe('FlowForge - Instance Snapshots', () => {
...
         cy.get('button[data-action="create-snapshot"]').click()
 
         cy.get('[data-el="dialog-create-snapshot"]').should('be.visible')
+        // eslint-disable-next-line cypress/require-data-selectors
         cy.get('.ff-dialog-header').contains('Create Snapshot')
```
Comment (joepavitt) [inline test/e2e/frontend/cypress/tests/instances/version-history/snapshots.spec.js:57]: `data-el="create-client-dialog"` exists on the dialog, this should be included here as a selector

File: test/e2e/frontend/cypress/tests/instances/version-history/snapshots.spec.js [hunk 5]
```diff
@@ -438,6 +442,7 @@ describe('FlowForge - Instance Snapshots', () => {
             cy.get('[data-el="platform-dialog"] .ff-btn--danger').click()
             cy.wait('@deleteSnapshot')
             if (count === 1) {
+                // eslint-disable-next-line cypress/require-data-selectors
                 cy.get('main').contains('Create your First Snapshot')
```
Comment (joepavitt) [inline test/e2e/frontend/cypress/tests/instances/version-history/snapshots.spec.js:446]:
```suggestion
                cy.get('div[data-el="empty-state"]').contains('Create your First Snapshot')
```

File: test/e2e/frontend/cypress/tests/instances/version-history/snapshots.spec.js [hunk 6]
```diff
@@ -723,15 +728,19 @@ describe('FlowForge shows audit logs', () => {
     })
 
     it('for when a snapshot is created', () => {
+        // eslint-disable-next-line cypress/require-data-selectors
```
Comment (joepavitt) [inline test/e2e/frontend/cypress/tests/instances/version-history/snapshots.spec.js:731]: and all of these, we should be using `data-` selectors, not classes

**Review by joepavitt — DISMISSED** [review]

File: test/e2e/frontend/cypress/tests/instances/version-history/snapshots.spec.js [hunk 7]
```diff
@@ -48,11 +49,14 @@ describe('FlowForge - Instance Snapshots', () => {
...
 
+        // eslint-disable-next-line cypress/require-data-selectors
         cy.get('button[data-action="create-snapshot"]').click()
 
         cy.get('[data-el="dialog-create-snapshot"]').should('be.visible')
+        // eslint-disable-next-line cypress/require-data-selectors
         cy.get('.ff-dialog-header').contains('Create Snapshot')
         // disabled primary button by default
+        // eslint-disable-next-line cypress/require-data-selectors
         cy.get('.ff-dialog-box button.ff-btn.ff-btn--primary').should('be.disabled')
```
[the remark this answers is not in the document]
Comment (cstns) [reply test/e2e/frontend/cypress/tests/instances/version-history/snapshots.spec.js:59]: this one doesn't have a data selector

[the remark this answers is not in the document]
Comment (cstns) [reply test/e2e/frontend/cypress/tests/instances/version-history/snapshots.spec.js:59]: my local linter is failing after their removal

**Comment by cstns** [comment]
I'll mark the e2e selector changes as resolved and take care of them on the follow up task

**Comment by cstns** [comment]
created https://github.com/FlowFuse/flowfuse/issues/4685 as a follow up task. Feel free to add things that I may have missed

**Review by cstns — APPROVED** [review]

**Pull request merged**